### DISCUSSION TITLE
Calling the functions of the config for once

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#550] [General]
Hello Experts,

I am trying to build the categorization for my analysis and it has turned out to be quite a heavy one. I have noticed that the categorization is being called before each and every task. Is it really necessary? 
I found one hack in `bbww` repo:

```py
def call_once_on_config(include_hash=False):
    """
    Parametrized decorator to ensure that function *func* is only called once for the config *config*
    """
    def outer(func):
        @wraps(func)
        def inner(config, *args, **kwargs):
            tag = f"{func.__name__}_called"
            if include_hash:
                tag += f"_{func.__hash__()}"

            if config.has_tag(tag):
                return

            config.add_tag(tag)
            return func(config, *args, **kwargs)
        return inner
    return outer
```

It probably targets the same, right?
If so, I checked that the categorization is being called every time ...
any idea would be very helpful

Thank you :)

**Comment by mafrahm** [comment]
Hi @gsaha009, the idea of this wrapper function is to call functions that modify our main config as part of the workflow (e.g. in some init function) only once. I implemented this because I want to add categories only at a specific point in my workflow (after calling a certain Producer). This "hack" is added because the init function of Producers are running quite often (e.g. to resolve different column requirements based on the used dataset/shift), but I only want to modify my config a single time.

This, however, all happens as part of resolving our task tree before the actual run methods of our tasks are started. The actual workload is usually passed to some other worker (on a different machine when submitting jobs) and therefore the setup of the config is repeated.

In theory, it should be possible to only create our central config a single time and then send that one to all other workers (e.g. via cloudpickle, although I never tried that myself). However, I am not sure how well this works when the config itself "evolves" based on which selectors/producers etc. have been requested.

I hope these comments are helpful to you. If you manage to find a way to minimize this configuration overhead, I'd be interested in seeing your solution.

**Reply by gsaha009** [reply]
Hi @mafrahm 
Thank you for your response. I think I get your point here. If I get some idea to minimize the config overhead, I will let you know. 

Gourab
